### ISSUE TITLE
Color management/generation and themes

### ISSUE THREAD

**Issue by JayPanoz** [edrlab/thorium-web#26]
Right now our primary color is #4d4d4d. To have a contrast that [passes WCAG AA for graphical objects and UI Components, it needs to be at least `60%` opacity](https://webaim.org/resources/contrastchecker/?fcolor=4D4D4Dc4&bcolor=FFFFFF), which makes the difference between normal and immersive modes kinda difficult to notice.

**Comment by HadrienGardeur** [comment]
I guess that the implementation of themes could also affect this value. Is there any way to dynamically calculate this contrast ratio as well?

**Comment by JayPanoz** [comment]
I know they have had issues specc’ing this in CSS as it’s a pretty popular issue but that means lots of options exist in JS/TS-land. Even [React actually](https://github.com/mirayatech/color-contrast-checker) – with an explainer in the ReadMe. 

I guess the main challenge may be building upon the known solutions to find the nearest match with good enough contrast in a dynamic manner, and reliably. But in theory it should be feasible.

**Comment by JayPanoz** [comment]
Took a deeper look at MUI, which has some world-class and extensive color manipulation utility to achieve some accessibility in their theming. 

- [utils can be found in colorManipulator](https://github.com/mui/material-ui/blob/817470c5a7c2ae721401e308b961005730cdf2ab/packages/mui-system/src/colorManipulator/colorManipulator.js) – I’m legit sad it doesn’t exist as a standalone package tbh, especially as there is a common approach across contrast utilities so you’re pretty much reimplementing the same functions everywhere;
- they have a `contrastThreshold` you can set in palette/themes but it’s just [a number for comparison](https://github.com/mui/material-ui/blob/817470c5a7c2ae721401e308b961005730cdf2ab/packages/mui-material/src/styles/createPalette.js#L206-L208) (if/else) in order to pick between light and dark text depending on the background color;
- they made it clear it’s the [same logic as Bootstrap in a comment](https://github.com/mui/material-ui/blob/817470c5a7c2ae721401e308b961005730cdf2ab/packages/mui-material/src/styles/createPalette.js#L201-L208);
- this means their theme provider won’t automagically pick a color for you;
- very few utilities go the extra mile to provide values that match the contrast criteria, but [they do exist](https://aristath.github.io/wcagColors.js/), albeit in a complex form.

**Comment by danielweck** [comment]
have you looked into https://www.radix-ui.com/colors ?

**Comment by JayPanoz** [comment]
@danielweck Not yet but thanks for the reminder as I’ve been looking into Libraries and how they manage theming in general   🙏

**Comment by JayPanoz** [comment]
For reference, [how colors are generated in Radix](https://github.com/radix-ui/website/blob/main/components/generateRadixColors.tsx). Couldn’t take an extensive look yet though, so can’t describe what is happening, I’m just logging the URL so that I remember.

**Comment by JayPanoz** [comment]
Rename the issue given the scope has changed, and will not be limited to immersive mode (and its `opacity`/`alpha`). 

In the future we may want to provide with automatic generation of palettes, etc. But also helpers that can be used when creating a theme, as can be seen in MUI for instance.

**Comment by JayPanoz** [comment]
After extending themes properties in the Playground preferences, I guess a sensible first implementation could be to create [tonal palettes](https://m3.material.io/styles/color/system/how-the-system-works#3ce9da92-a118-4692-8b2c-c5c52a413fa6) from main colors e.g. `background` can be used for `hover` for instance, `text` for `subdue` or `disable` etc. and pick the first color in the tonal palette that meets a given/preset contrast ratio. 

In that case it could either be automatic when a theme property is `null` or opt-in (an helper function). Results are not going to be necessarily perfect every time but that should at least help get a starting point you can fine-tune if you're not happy with the result. 

That said, I’d be happy to hear how others would approach this issue.

**Comment by JayPanoz** [comment]
[Related discussion in Thorium](https://github.com/edrlab/thorium-reader/discussions/2817#discussioncomment-12292945)

It looks like the scope of color helpers could be larger than ReadiumCSS and could provide benefits to multiple projects.

**Comment by JayPanoz** [comment]
As a heads up, this issue will probably be moved and handled elsewhere, and not directly in Playground as it makes sense to have it as a helper you can use across projects.
